lib/hls/hls_parser.js

```diff
@@ -1966,6 +2035,53 @@ shaka.hls.HlsParser = class {
       references.push(reference);
     }
 
+    // If some segments have sync times, but not all, extrapolate the sync
+    // times of the ones with none.
+    const someSyncTime = references.some((ref) => ref.syncTime != null);
+    if (someSyncTime) {
+      for (let i = 0; i < references.length; i++) {
+        const reference = references[i];
+        if (reference.syncTime != null) {
+          // No need to extrapolate.
+          continue;
+        }
+        // Find the nearest segment with syncTime, in either direction.
+        // This looks forward and backward simultaneously, keeping track of what
+        // to offset the syncTime it finds by as it goes.
+        let forwardAdd = 0;
+        let backwardAdd = 0;
+        for (let j = 0; reference.syncTime == null; j++) {
+          for (const forward of [false, true]) {
+            const other = references[i + j * (forward ? 1 : -1)];
+            if (other) {
+              if (other.syncTime != null) {
+                reference.syncTime = other.syncTime;
+                reference.syncTime += forward ? forwardAdd : backwardAdd;
+                break;
+              }
+              if (forward) {
+                forwardAdd -= other.endTime - other.startTime;
+              }
+              if (!forward) {
+                backwardAdd += other.endTime - other.startTime;
+              }
```

Comment:
I think you're off-by-one here somewhere.

If reference i has no sync time, and i+1 does, the difference between the sync times should be the duration of i.  If reference i-1 is the first match, the difference between the sync times should be the duration of i-1.

Since the loop starts at j=0, reference == other in the first pass.  Say the durations of the segments going forward (including j=0) are d0, d1, d2, ... and the durations of the segments going backward (including j=0) are e0, e1, e2, ..., then:

|matching j|correct offset|
|---|---|
|0|impossible|
|+1|d0|
|+2|d0+d1|
|+3|d0+d1+d2|
|+4|d0+d1+d2+d3|
|-1|e1|
|-2|e1+e2|
|-3|e1+e2+e3|
|-4|e1+e2+e3+e4|

The backward offsets should never include the current segment duration (j=0), but in this algorithm, they do.  Your answer would still be correct any time the segments have a consistent duration, though.

It took me a very long time to analyze this code, though.  I think this needs to be rewritten in a simpler way if possible, rather than adding another clause or exception.

I think we should also make sure we add test cases with varying segment lengths to catch issues like this.

Reply:
I reorganized this some. Do you think it is more readable this way? Or should I come up with an entirely different algorithm?